addons/hr_holidays/models/hr_leave.py

```diff
@@ -97,6 +97,114 @@ def _default_employee(self):
         ('date_check', "CHECK ( number_of_days_temp >= 0 )", "The number of days must be greater than 0."),
     ]
 
+    # Those fields are mostly for the interface only
+
+    date_from_date = fields.Date()
+    date_to_date = fields.Date()
+
+    method = fields.Selection(related='holiday_status_id.method', store=True, readonly=True)
+
+    date_from_am_pm = fields.Selection([('am', 'Morning'), ('pm', 'Afternoon')], default='am')
+    date_to_am_pm = fields.Selection([('am', 'Morning'), ('pm', 'Afternoon')], default='pm')
+
+    unit = fields.Selection([('half', 'Half Day'),
+                             ('day', '1 Day'),
+                             ('period', 'Period')],
+                            default='day')
+    unit_day = fields.Selection([('day', '1 Day'),
+                                 ('period', 'Period')],
+                                default='day')
+
+    number_of_hours = fields.Float(
+        'Hours Allocation', copy=False, readonly=True, compute='_compute_number_of_hours',
+        states={'draft': [('readonly', False)], 'confirm': [('readonly', False)]},
+        help='Number of hours of the leave request according to your working schedule.')
+
+    def _calc_days_temp(self):
+        is_set = self.date_from and self.date_to
+        is_set_date = self.date_from_date and self.date_to_date
+
+        if is_set and is_set_date:
+            date_from = False
+            date_to = False
+
+            Attendances = self.env['resource.calendar.attendance']
+            domain = [('calendar_id', '=', self.employee_id.resource_calendar_id.id)]
+
+            first_day = fields.Date.from_string(self.date_from_date)
+            last_day = fields.Date.from_string(self.date_to_date)
+
+            attendance_from = Attendances.search(domain + [('dayofweek', '=', first_day.weekday()), ('day_period', '=', 'morning'), ('calendar_id', '=', self.employee_id.resource_calendar_id.id)], limit=1)
+            attendance_to = Attendances.search(domain + [('dayofweek', '=', last_day.weekday()), ('day_period', '=', 'afternoon'), ('calendar_id', '=', self.employee_id.resource_calendar_id.id)], limit=1)
+
+            # In case we only have an attendance in the morning or only in the afternoon
+            if not attendance_to:
+                attendance_to = attendance_from
+
+            if not attendance_from:
+                attendance_from = attendance_to
+
+            if self.unit == 'day' or (self.unit == 'period' and self.method == 'day'):
```

Comment:
There is some strange code here. I will try to explain, please forgive me if I tell stupid things, dates are complicated :p . My point is about computing the hour_from and hour_to.
 * It seems to me that working in half day should be straightforward: we should find one attendance and take its from_to
 * Working in day seems ok
 * Working in hours seems ok

I think the code could be simpler, but it require a bit in-depth inspection to be sure :) .